File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/UpdateOrderUseCase.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.domain.models.Order;
import com.example.orderservice.domain.models.OrderItem;
import com.example.orderservice.domain.ports.OrderRepository;
import com.example.orderservice.domain.ports.CacheManager;
import com.example.orderservice.domain.order_id.OrderId;
import org.springframework.stereotype.Component;

import java.util.List;

/**
 * Update Order Use Case demonstrating Write-Through pattern.
 * 
 * Pattern:
 * 1. Update database
 * 2. Immediately update cache (write-through)
 * 3. Return updated order
 * 
 * This ensures cache consistency with database.
 */
@Component
public class UpdateOrderUseCase {
    
    private static final String CACHE_KEY_PREFIX = "order-service:order:";
    private static final String CACHE_KEY_SUFFIX = ":full";
    
    private final OrderRepository orderRepository;
    private final CacheManager cacheManager;
    
    public UpdateOrderUseCase(OrderRepository orderRepository, CacheManager cacheManager) {
        this.orderRepository = orderRepository;
        this.cacheManager = cacheManager;
    }
    
    /**
     * Update order with write-through caching.
     * 
     * @param orderId Order ID
     * @param items Updated order items
     * @return Updated order
     * @throws OrderNotFoundException if order doesn't exist
     */
    public Order execute(OrderId orderId, List<OrderItem> items) {
        // Step 1: Update database
        Order updatedOrder = orderRepository.update(orderId, items);
        
        // Step 2: Write-through - update cache immediately
        String cacheKey = buildCacheKey(orderId);
        cacheManager.put(cacheKey, updatedOrder);
        
        // Step 3: Return updated order
        return updatedOrder;
    }
    
    /**
     * Cancel order with cache invalidation.
     * 
     * @param orderId Order ID to cancel
     */
    public void cancelOrder(OrderId orderId) {
        // Update database
        orderRepository.cancel(orderId);
        
        // Invalidate cache (don't cache cancelled orders)
        invalidateCache(orderId);
    }
    
    /**
     * Invalidate order cache.
     * 
     * @param orderId Order ID to invalidate
     */
    public void invalidateCache(OrderId orderId) {
        String cacheKey = buildCacheKey(orderId);
        cacheManager.evict(cacheKey);
    }
    
    private String buildCacheKey(OrderId orderId) {
        return CACHE_KEY_PREFIX + orderId.getValue() + CACHE_KEY_SUFFIX;
    }
    
    public static class OrderNotFoundException extends RuntimeException {
        public OrderNotFoundException(OrderId orderId) {
            super("Order not found: " + orderId.getValue());
        }
    }
}

File: tools/cookiecutter-python/{{cookiecutter.project_slug}}/src/application/usecases/update_order_use_case.py
"""
Update Order Use Case demonstrating Write-Through pattern.

Pattern:
1. Update database
2. Immediately update cache (write-through)
3. Return updated order

This ensures cache consistency with database.
"""

from typing import List
from domain.models.order import Order
from domain.models.order_item import OrderItem
from domain.ports.order_repository import OrderRepository
from domain.ports.cache_manager import CacheManager
from domain.order_id import OrderId


class UpdateOrderUseCase:
    """Update order with write-through caching."""
    
    CACHE_KEY_PREFIX = "{{ cookiecutter.project_slug }}:order:"
    CACHE_KEY_SUFFIX = ":full"
    
    def __init__(self, order_repository: OrderRepository, cache_manager: CacheManager):
        """
        Initialize use case.
        
        Args:
            order_repository: Order repository port
            cache_manager: Cache manager port
        """
        self.order_repository = order_repository
        self.cache_manager = cache_manager
    
    def execute(self, order_id: OrderId, items: List[OrderItem]) -> Order:
        """
        Update order with write-through caching.
        
        Args:
            order_id: Order ID
            items: Updated order items
            
        Returns:
            Updated order
            
        Raises:
            OrderNotFoundException: If order doesn't exist
        """
        # Step 1: Update database
        updated_order = self.order_repository.update(order_id, items)
        if updated_order is None:
            raise OrderNotFoundException(order_id)
        
        # Step 2: Write-through - update cache immediately
        cache_key = self._build_cache_key(order_id)
        self.cache_manager.put(cache_key, updated_order)
        
        return updated_order
    
    def cancel_order(self, order_id: OrderId) -> None:
        """
        Cancel order with cache invalidation.
        
        Args:
            order_id: Order ID to cancel
        """
        # Update database
        self.order_repository.cancel(order_id)
        
        # Invalidate cache (don't cache cancelled orders)
        self.invalidate_cache(order_id)
    
    def invalidate_cache(self, order_id: OrderId) -> None:
        """
        Invalidate order cache.
        
        Args:
            order_id: Order ID to invalidate
        """
        cache_key = self._build_cache_key(order_id)
        self.cache_manager.evict(cache_key)
    
    def _build_cache_key(self, order_id: OrderId) -> str:
        """Build cache key from order ID."""
        return f"{self.CACHE_KEY_PREFIX}{order_id.value}{self.CACHE_KEY_SUFFIX}"


class OrderNotFoundException(Exception):
    """Raised when order is not found."""
    
    def __init__(self, order_id: OrderId):
        super().__init__(f"Order not found: {order_id.value}")
